crates/turborepo-lib/src/process/mod.rs

```diff
@@ -0,0 +1,286 @@
+//! `process`
+//!
+//! This module contains the code that is responsible for running the commands
+//! that are queued by run. It consists of a set of child processes that are
+//! spawned and managed by the manager. The manager is responsible for
+//! running these processes to completion, forwarding signals, and closing
+//! them when the manager is closed.
+//!
+//! As of now, the manager will execute futures in a random order, and
+//! must be either `wait`ed on or `stop`ped to drive state.
+
+mod child;
+
+use std::{
+    sync::{atomic::AtomicBool, Arc, Mutex},
+    time::Duration,
+};
+
+pub use child::Command;
+use futures::Future;
+use tokio::task::JoinSet;
+use tracing::{debug, trace};
+
+use self::child::{Child, ChildExit};
+
+/// A process manager that is responsible for spawning and managing child
+/// processes. When the manager is Open, new child processes can be spawned
+/// using `spawn`. When the manager is Closed, all currently-running children
+/// will be closed, and no new children can be spawned.
+#[derive(Debug, Clone)]
+pub struct ProcessManager {
+    is_closing: Arc<AtomicBool>,
+    children: Arc<Mutex<Vec<child::Child>>>,
+}
+
+impl ProcessManager {
+    pub fn new() -> Self {
+        ProcessManager {
+            is_closing: Arc::new(AtomicBool::new(false)),
+            children: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+}
+
+impl ProcessManager {
+    /// Spawn a new child process to run the given command.
+    ///
+    /// The handle of the child can be either waited or stopped by the caller,
+    /// as well as the entire process manager.
+    ///
+    /// If spawn returns None,
+    pub fn spawn(&self, command: child::Command, stop_timeout: Duration) -> Option<child::Child> {
+        if self.is_closing.load(std::sync::atomic::Ordering::Relaxed) {
+            return None;
+        }
```

Comment:
I'm think we might have a TOCTOU race condition here that could result in unmanaged children:
 - A call to `spawn` and `close` happen at the same time
 - The `spawn` call checks `is_closing`, but the setting of `true` hasn't happened from `close`
 - The `close` call then sets the `is_closing`, it then proceeds to lock children and send closing signals to all children. It then releases the mutex
 - The `spawn` call, waiting on the lock held by `close` can now proceed to add the newly spawned `child` into the children vector
 - After all of the children close `close` proceeds and clears the vector of children since it believes it has closed them all. This leaves us with no handle for the `Child` spawned by the `spawn` call.

Reply:
Nice, I think I fixed it by requiring a lock of the child vec before checking if we are closing. I really want to keep this to support 'kill promotion' allowing us to stop / kill things at the process manager level. If we lock the whole thing, we can't send new commands while waiting for another to complete.